### ISSUE TITLE
##WARNING##:PowerShell Core (pwsh.exe) is not available on agent machine. Falling back to using Windows PowerShell (powershell.exe). This can cause reduced performance

### ISSUE THREAD

**Issue by penihel** [microsoft/azure-pipelines-tasks#14595]
I have a pipeline with that i call the "IIS web app deploy" task twice.

the first i got error, but the last, its ok.

the first log
```
2021-03-16T14:45:15.8461138Z ==============================================================================
2021-03-16T14:45:15.8461433Z Task         : IIS web app deploy
2021-03-16T14:45:15.8461693Z Description  : Deploy a website or web application using Web Deploy
2021-03-16T14:45:15.8461944Z Version      : 0.184.0
2021-03-16T14:45:15.8462138Z Author       : Microsoft Corporation
2021-03-16T14:45:15.8462483Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/deploy/iis-web-app-deployment-on-machine-group
2021-03-16T14:45:15.8462889Z ==============================================================================
2021-03-16T14:45:16.6126018Z ##WARNING##:PowerShell Core (pwsh.exe) is not available on agent machine. Falling back to using Windows PowerShell (powershell.exe). This can cause reduced performance. Please install the newer version of PowerShell for improved performance.
2021-03-16T14:45:16.6165708Z [command]C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe -NoLogo -NoProfile -NonInteractive -Command "Expand-Archive -Path \"C:\azagent\A2\_work\r1\a\IMPLANTANET\Common\Implanta.Common.Web.CDN.zip\" -DestinationPath \"C:\azagent\A2\_work\_temp\temp_web_package_11780163432773327\" -Force"
2021-03-16T14:45:34.3159510Z Remove-Item : Cannot find path 'C:\azagent\A2\_work\_temp\temp_web_package_11780163432773327\Content\D_C\a\1\s\src\Comm
2021-03-16T14:45:34.3160324Z on\Implanta.Common.Web.CDN\obj\Release\Package\PackageTmp\BarCodeHandler.ashx' because it does not exist.
2021-03-16T14:45:34.3169917Z At 
2021-03-16T14:45:34.3170466Z C:\windows\system32\WindowsPowerShell\v1.0\Modules\Microsoft.PowerShell.Archive\Microsoft.PowerShell.Archive.psm1:411 
2021-03-16T14:45:34.3171025Z char:46
2021-03-16T14:45:34.3171407Z + ...                 $expandedItems | % { Remove-Item $_ -Force -Recurse }
2021-03-16T14:45:34.3171875Z +                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-03-16T14:45:34.3172488Z     + CategoryInfo          : ObjectNotFound: (C:\azagent\A2\_...odeHandler.ashx:String) [Remove-Item], ItemNotFoundEx 
2021-03-16T14:45:34.3172936Z    ception
2021-03-16T14:45:34.3173273Z     + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.RemoveItemCommand
2021-03-16T14:45:34.3173584Z  
2021-03-16T14:45:34.3196734Z Remove-Item : Cannot find path 'C:\azagent\A2\_work\_temp\temp_web_package_11780163432773327\Content\D_C\a\1\s\src\Comm
2021-03-16T14:45:34.3197465Z on\Implanta.Common.Web.CDN\obj\Release\Package\PackageTmp\bin\' because it does not exist.
```

but in the same pipeline, the next task, runs ok.

the second log

```
2021-03-16T14:46:13.4512632Z ##[section]Starting: Deploy IIS Website/App: CRM
2021-03-16T14:46:13.4593875Z ==============================================================================
2021-03-16T14:46:13.4594171Z Task         : IIS web app deploy
2021-03-16T14:46:13.4594441Z Description  : Deploy a website or web application using Web Deploy
2021-03-16T14:46:13.4594682Z Version      : 0.184.0
2021-03-16T14:46:13.4594888Z Author       : Microsoft Corporation
2021-03-16T14:46:13.4595225Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/deploy/iis-web-app-deployment-on-machine-group
2021-03-16T14:46:13.4595626Z ==============================================================================
2021-03-16T14:46:17.9437651Z ##[warning]Unable to apply transformation for the given package. Verify the following.
2021-03-16T14:46:17.9439940Z ##[warning]1. Whether the Transformation is already applied for the MSBuild generated package during build. If yes, remove the <DependentUpon> tag for each config in the csproj file and rebuild. 
2021-03-16T14:46:17.9441945Z ##[warning]2. Ensure that the config file and transformation files are present in the same folder inside the package.
2021-03-16T14:46:18.0400057Z Initiated variable substitution in config file : C:\azagent\A2\_work\_temp\temp_web_package_04240053675901212\web.config
2021-03-16T14:46:18.0450274Z Skipped Updating file: C:\azagent\A2\_work\_temp\temp_web_package_04240053675901212\web.config
2021-03-16T14:46:18.0450803Z XML variable substitution applied successfully.
2021-03-16T14:46:18.0906546Z [command]"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:IisApp='C:\azagent\A2\_work\_temp\temp_web_package_04240053675901212' -dest:iisApp='daf-07.implantadev.net.br/crm' -skip:Directory=App_Data -skip:Directory=^api$
2021-03-16T14:46:18.4921467Z Info: Object dirPath (daf-07.implantadev.net.br/crm\api) skipped due to skip directive 'CommandLineSkipDirective 2'.
2021-03-16T14:46:18.5072541Z Info: Adding file (daf-07.implantadev.net.br/crm\1-es2015.e18d59dcb22750d8ccff.js).
2021-03-16T14:46:18.5134447Z Info: Adding file (daf-07.implantadev.net.br/crm\1-es5.e18d59dcb22750d8ccff.js).
2021-03-16T14:46:18.5139507Z Info: Adding file (daf-07.implantadev.net.br/crm\10-es2015.5b40fe0727b1a8ff0c42.js).
2021-03-16T14:46:18.5146813Z Info: Adding file (daf-07.implantadev.net.br/crm\10-es5.5b40fe0727b1a8ff0c42.js).
2021-03-16T14:46:18.5154719Z Info: Adding file (daf-07.implantadev.net.br/crm\11-es2015.8c7132ed62bcc7eed110.js).
2021-03-16T14:46:18.5163928Z Info: Adding file (daf-07.implantadev.net.br/crm\11-es5.8c7132ed62bcc7eed110.js).
```

## Required Information

**Enter Task Name**: IIS web app deploy

## Environment
- Server - Azure Pipelines runing o VM o Azure

Accountname:https://dev.azure.com/implantainformatica/
ProjectName:IMPLANTANET
Release definition name:DAF
Link: https://dev.azure.com/implantainformatica/IMPLANTANET/_releaseProgress?releaseId=40568&_a=release-environment-deployment-group-logs&environmentId=880539&deploymentGroupPhaseId=106117&agentName=DEV-DAF-1

- Agent : Private: 
     -  OS: Windows
     - Agent version :2.181.1

## Issue Description

the first return warning, and due to the alert (I believe) the copy of the files fails, you cannot extract the files correctly from the "zip". 

When this error occurs, the target folder (iis application) looks like that:

![image](https://user-images.githubusercontent.com/1167747/111332185-549ccd00-8650-11eb-9f3e-324a504dfcb8.png)
but with all folders without files

**Comment by penihel** [comment]
I noticed, that Saturday worked

```
2021-03-13T04:07:06.0259138Z ##[section]Starting: Deploy CDN
2021-03-13T04:07:06.0367340Z ==============================================================================
2021-03-13T04:07:06.0367787Z Task         : IIS web app deploy
2021-03-13T04:07:06.0368315Z Description  : Deploy a website or web application using Web Deploy
2021-03-13T04:07:06.0368685Z Version      : 0.179.0
2021-03-13T04:07:06.0368992Z Author       : Microsoft Corporation
2021-03-13T04:07:06.0369501Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/deploy/iis-web-app-deployment-on-machine-group
2021-03-13T04:07:06.0370051Z ==============================================================================
2021-03-13T04:07:06.4297490Z [command]C:\azagent\A1\_work\_tasks\IISWebAppDeploymentOnMachineGroup_1b467810-6725-4b6d-accd-886174c09bba\0.179.0\node_modules\webdeployment-common-v2\7zip\7z.exe x -oC:\azagent\A1\_work\_temp\temp_web_package_16331721703286406 C:\azagent\A1\_work\r1\a\IMPLANTANET\Common\Implanta.Common.Web.CDN.zip
2021-03-13T04:07:06.4624132Z 
2021-03-13T04:07:06.4637351Z 7-Zip [64] 16.00 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-10
2021-03-13T04:07:06.4641076Z 
2021-03-13T04:07:06.4642111Z Scanning the drive for archives:
2021-03-13T04:07:06.4647875Z 1 file, 13024068 bytes (13 MiB)
2021-03-13T04:07:06.4651069Z 
2021-03-13T04:07:06.4654675Z Extracting archive: C:\azagent\A1\_work\r1\a\IMPLANTANET\Common\Implanta.Common.Web.CDN.zip
2021-03-13T04:07:08.8024050Z --
2021-03-13T04:07:08.8024845Z Path = C:\azagent\A1\_work\r1\a\IMPLANTANET\Common\Implanta.Common.Web.CDN.zip
2021-03-13T04:07:08.8025321Z Type = zip
2021-03-13T04:07:08.8025695Z Physical Size = 13024068
2021-03-13T04:07:08.8025980Z 
2021-03-13T04:07:08.8027840Z Everything is Ok
2021-03-13T04:07:08.8028176Z 
2021-03-13T04:07:08.8028477Z Folders: 198
2021-03-13T04:07:08.8028779Z Files: 1184
2021-03-13T04:07:08.8029287Z Size:       45900419
2021-03-13T04:07:08.8029631Z Compressed: 13024068
```

But, today not
```
2021-03-16T12:01:37.1473166Z ##[section]Starting: Deploy CDN
2021-03-16T12:01:37.1574539Z ==============================================================================
2021-03-16T12:01:37.1574862Z Task         : IIS web app deploy
2021-03-16T12:01:37.1575127Z Description  : Deploy a website or web application using Web Deploy
2021-03-16T12:01:37.1575382Z Version      : 0.184.0
2021-03-16T12:01:37.1575579Z Author       : Microsoft Corporation
2021-03-16T12:01:37.1575927Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/deploy/iis-web-app-deployment-on-machine-group
2021-03-16T12:01:37.1576445Z ==============================================================================
2021-03-16T12:01:37.9633541Z ##WARNING##:PowerShell Core (pwsh.exe) is not available on agent machine. Falling back to using Windows PowerShell (powershell.exe). This can cause reduced performance. Please install the newer version of PowerShell for improved performance.
2021-03-16T12:01:37.9676786Z [command]C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe -NoLogo -NoProfile -NonInteractive -Command "Expand-Archive -Path \"C:\azagent\A1\_work\r1\a\IMPLANTANET\Common\Implanta.Common.Web.CDN.zip\" -DestinationPath \"C:\azagent\A1\_work\_temp\temp_web_package_2484840271019726\" -Force"
2021-03-16T12:01:58.4396150Z Remove-Item : Cannot find path 'C:\azagent\A1\_work\_temp\temp_web_package_2484840271019726\Content\D_C\a\1\s\src\Commo
2021-03-16T12:01:58.4396783Z n\Implanta.Common.Web.CDN\obj\Release\Package\PackageTmp\BarCodeHandler.ashx' because it does not exist.
2021-03-16T12:01:58.4397094Z At 
2021-03-16T12:01:58.4407658Z C:\windows\system32\WindowsPowerShell\v1.0\Modules\Microsoft.PowerShell.Archive\Microsoft.PowerShell.Archive.psm1:411 
2021-03-16T12:01:58.4408301Z char:46
2021-03-16T12:01:58.4408593Z + ...                 $expandedItems | % { Remove-Item $_ -Force -Recurse }
2021-03-16T12:01:58.4409021Z +                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-03-16T12:01:58.4409476Z     + CategoryInfo          : ObjectNotFound: (C:\azagent\A1\_...odeHandler.ashx:String) [Remove-Item], ItemNotFoundEx 
2021-03-16T12:01:58.4409835Z    ception
2021-03-16T12:01:58.4410104Z     + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.RemoveItemCommand
2021-03-16T12:01:58.4410348Z  
2021-03-16T12:01:58.4435895Z Remove-Item : Cannot find path 'C:\azagent\A1\_work\_temp\temp_web_package_2484840271019726\Content\D_C\a\1\s\src\Commo
2021-03-16T12:01:58.4436429Z n\Implanta.Common.Web.CDN\obj\Release\Package\PackageTmp\bin\' because it does not exist.
2021-03-16T12:01:58.4436708Z At 
2021-03-16T12:01:58.4437061Z C:\windows\system32\WindowsPowerShell\v1.0\Modules\Microsoft.PowerShell.Archive\Microsoft.PowerShell.Archive.psm1:411 
2021-03-16T12:01:58.4437399Z char:46
```

Just the task version was changed (0.179.0 to 0.184.0)

**Comment by penihel** [comment]
Im pretty sure that this is the problem https://github.com/microsoft/azure-pipelines-tasks/commit/92ae148a02b1a174cc212222f9d92b07dc536ea6

**Comment by pixelnix** [comment]
This change to use Expand-Archive instead of 7zip broke the IIS web app deploy task on all my servers. Please fix or revert it.

**Comment by penihel** [comment]
The problem continues,

but I managed to get around it by installing PowerShell Core 7.1 on my server

```powershell
# download the installation code
$code = Invoke-RestMethod -Uri https://aka.ms/install-powershell.ps1
# turn it into a function
$null = New-Item -Path function:Install-PowerShell -Value $code
# run the function
Install-PowerShell -UseMSI -Quiet
```

So, i restarted the server and ran another deploy, and finally works fine!

**Comment by pixelnix** [comment]
After reading the commit referenced above, it seems the developers left a switch to force the IIS Web App Deploy task back to using 7zip for artifact extraction. Just add a pipeline variable "ADO_FORCE_USE_7ZIP" set to "true".

**Comment by pixelnix** [comment]
For reference:
https://github.com/microsoft/azure-pipelines-tasks/commit/92ae148a02b1a174cc212222f9d92b07dc536ea6#diff-46a33d08a56b1f9b206377aa6c90dccb24740d1a652f0b425b9c789d302d4aaeR14

**Comment by chshrikh** [comment]
https://github.com/microsoft/azure-pipelines-tasks/issues/14607

**Comment by GeorgePotterCivica** [comment]
We've been having the same issues as described here: https://github.com/microsoft/azure-pipelines-tasks/issues/14595#issuecomment-800602330

Previously this task was using 7zip for the deployment, now as of this update it always tries to use Powershell Core without ensuring that it's installed on the machine. The fallback behaviour of using Windows Powershell is broken because the expand-archive command does not exist in Windows Powershell.

In short, if you **must** do updates that override existing functionality (e.g. stopping the use of 7zip) please could you at least make sure that the fallback for your new approach actually works.

File extraction commands exist in Windows Powershell so it shouldn't be difficult to substitute those for expand-archive if the task falls back to using Windows Powershell.

**Comment by pixelnix** [comment]
Also, do these tasks not have incrementable version numbers? Azure tells me that the IIS Web App Deploy task is version 0, with no other versions available. If this change doesn't warrant a version bump, retaining the old version that doesn't use Expand-Archive, I'm not sure what does.

**Comment by pixelnix** [comment]
@chshrikh The notes you wrote in #14607 also don't appear in the README for this task (https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/IISWebAppDeploymentOnMachineGroupV0/README.md). If the plan is really to leave things as they stand now, that documentation should probably be updated to say what version of PowerShell is required for the task to function normally, and how to force it to revert back to using 7zip.

**Comment by chshrikh** [comment]
@GeorgePotterCivica can you try your scenario by adding pipeline variable named - ADO_FORCE_USE_7ZIP and  set its value as true. With this task will forcefully use the 7Zip for unzipping

**Comment by chshrikh** [comment]
@pixelnix we will update the read md once our new solution reaches to all customers.

**Comment by GeorgePotterCivica** [comment]
> @GeorgePotterCivica can you try your scenario by adding pipeline variable named - ADO_FORCE_USE_7ZIP and set its value as true. With this task will forcefully use the 7Zip for unzipping

That didn't work for us, but installing PowerShell Core did, so we're happy with the new setup now that it's working again. Thank you :)

**Comment by ebcdavid** [comment]
I can verify that the pipeline variable solved the same issue for us; however, we are running many locally hosted agents that will need powershell approved/installed,  or we have many release pipelines that will need an edit.  Will there be a fix for this issue so this isn't work we have to do, or is it a permanent (breaking) change?  CI/CD is broken for us at the moment.

**Comment by pixelnix** [comment]
The thing that the developer called "experimental" in the PR is now the official accepted way to back out the breaking change to this task. Good times.

**Comment by pixelnix** [comment]
I'll also note that our servers are Azure-hosted windows server instances that were created as recently as December, and they don't have the requisite version of Powershell to run this task. Seems like this change could easily cause problems for far more users than for whom it fixes the issue with UTF8 characters in filenames.

**Comment by carterlist** [comment]
I have the same situation as ebcdavid- CI/CD is broken at the moment.

**Comment by anil-khatter** [comment]
I have the exact same situation as @ebcdavid. As a matter of fact, we have to cancel our Production deployment because we saw this issue surface all of a sudden. Do we have a timeline for a permanent fix for this? Also, if we could tag such breaking changes to a different major version so as to be backward compatible. I didn't find any way to specify the old working version in release pipelines. Thanks.

**Comment by lukepat** [comment]
We had the same issue, installing powershell core 7.1 as shown by penihel resolved the issue, thankfully we didn't have many servers to apply the workaround to.

**Comment by irontoby** [comment]
This is a terrible way to introduce a breaking change. The version number of the Task should have been increased so we could at least choose the old behavior. We spent 2 hours last night trying to figure out why all our release pipelines started failing before finally finding the obscure `ADO_FORCE_USE_7ZIP` setting.

The logic to detect whether to use PowerShell or 7Zip should at least check to see whether `Expand-Archive` exists in PowerShell and revert to 7Zip if not. If that's not possible we at least need a way to change the setting more easily.

**Comment by shawnbeeson** [comment]
This also broke all of our release pipelines and has delayed a production release.  Are you really not going to roll back this change?  Installing the powershell core requires taking production machines offline and scheduling a maintenance window (causing further delays).  

I would simply like a answer as to whether or not this is going to be fixed so the the IIS Web App Deploy activity functions properly as it did before?

**Comment by shawnbeeson** [comment]
> @pixelnix we will update the read md once our new solution reaches to all customers.

Should we all take this to mean that you will NOT be rolling this back?

**Comment by GeorgePotterCivica** [comment]
I do like how the status of this ticket is clearly being updated but no one's bothered to answer any of the questions yet.

Surely this isn't complex to fix? Either roll back the change **or** issue some workaround guidance **or** do a patch to have it default back to 7zip if Powershell Core isn't installed.

**Comment by mattsteinRRD** [comment]
This is affecting us now too.  We have hundreds of servers.  The log says that it is falling back to Powershell but the extract does not actually work.  Please fix this quickly!

**Comment by pixelnix** [comment]
Official responses notwithstanding, there appears to be a PR to back this out by reversing the sense of the change - keeping PowerShell as a method of archive extraction but making 7zip the default.
https://github.com/microsoft/azure-pipelines-tasks/pull/14626

**Comment by shawnbeeson** [comment]
> Official responses notwithstanding, there appears to be a PR to back this out by reversing the sense of the change - keeping PowerShell as a method of archive extraction but making 7zip the default.
> #14626
> 
> -edit- please note I am not speaking for the Azure devs here

Looking at #14626 is still appears that they are going to force/require that we add a pipeline variable: ADO_FORCE_USE_PSUNZIP in order for the current IIS Web App Deploy step to work properly....Complete hack IMO.  Its going to require Admins to update every pipeline currently that currently uses the IIS Web App Deploy action.

Also, It appears the PR was merged into master.  Can an Op please respond with an ETA, when this hack will be in Production?

**Comment by pixelnix** [comment]
> Looking at #14626 is still appears that they are going to force/require that we add a pipeline variable: ADO_FORCE_USE_PSUNZIP in order for the current IIS Web App Deploy step to work properly....Complete hack IMO. Its going to require Admins to update every pipeline currently that currently uses the IIS Web App Deploy action.
> 
> Also, It appears the PR was merged into master. Can an Op please respond with an ETA, when this hack will be in Production?

FWIW I read that PR differently; I think all your pipelines that do not declare any hacky variables will work as they did before. Only if a pipeline is run with ADO_FORCE_USE_PSUNZIP=true will the extraction task use PowerShell, otherwise it'll use 7zip as it did before the breaking change. Users who are experiencing the original issue Azure devs tried to fix (UTF characters in filenames) will need to add that variable to _their_ pipelines in order for them to use PowerShell Expand-Archive. Hoping I'm right here.

**Comment by Charlove** [comment]
Below are the coordinates that helped me get around this problem

  - I installed the version PowerShell 7.1.3
  - I accessed Windows services (services.msc)
  - I restarted Azure services
 
When you restart Azure services on Windows, the settings are updated.
 
Hope this helps

**Comment by chshrikh** [comment]
@pixelnix you are correct

**Comment by chshrikh** [comment]
Sorry for the inconvenience caused.
We are working on fix which will solve both the issues 1) defaulting to existing 7 zip tooling and 2) Handling the special characters with msbuild packages. 
Only customers who are facing issues with special character in msbuild need to update their pipelines and add a special variable "ADO_FORCE_USE_PSUNZIP" with value 'true', other customer does not need to update their pipeline, their scenario should work fine as it was working till last week. Even for customer who has added pipeline variable "ADO_FORCE_USE_7ZIP" or installed PS core, there should not be any change required, their scenario will continue to work without any pipeline / task update.
We are already in process of rolling out the new fix with 0.184.1 version of task. It will reach to all customer within next three days.
Thanks for your support

**Comment by chshrikh** [comment]
The deployment of 0.184.1 has been completed, as stated with previous comment, existing pipeline should was as expected and defaulted to use 7 zip
